src/ActionMenu.tsx

```diff
@@ -0,0 +1,132 @@
+  const anchorRef = useRef<HTMLElement>(null)
+  const overlayRef = useRef<HTMLDivElement>(null)
+
+  const anchorId = `dropdownMenuAnchor-${uniqueId()}`
+  const [open, setOpen] = useState<boolean>(false)
+  const [state, setState] = useState<'closed' | 'buttonFocus' | 'listFocus'>('closed')
+  const onDismiss = useCallback(() => {
+    setOpen(false)
+    setState('closed')
+  }, [])
+
+  const onAnchorKeyDown = useCallback(
+    (event: React.KeyboardEvent<HTMLElement>) => {
+      if (!event.defaultPrevented) {
+        if (state === 'closed') {
+          if (event.key === 'ArrowDown' || event.key === 'ArrowUp') {
+            setState('listFocus')
+            setOpen(true)
+            event.preventDefault()
+          } else if (event.key === ' ' || event.key === 'Enter') {
+            setState('buttonFocus')
+            setOpen(true)
+            event.preventDefault()
+          }
+        } else if (state === 'buttonFocus') {
+          if (['ArrowDown', 'ArrowUp', 'Tab', 'Enter'].indexOf(event.key) !== -1) {
+            setState('listFocus')
+            event.preventDefault()
+          } else if (event.key === 'Escape') {
+            setState('closed')
+            onDismiss()
+            event.preventDefault()
+          }
+        }
+      }
+    },
+    [state, onDismiss]
+  )
+  const onAnchorClick = useCallback(
+    (event: React.MouseEvent<HTMLElement>) => {
+      if (!event.defaultPrevented && event.button === 0 && !open) {
+        setOpen(true)
+        setState('buttonFocus')
+      }
+    },
+    [open]
+  )
+
+  const {position} = useAnchoredPosition({anchorElementRef: anchorRef, floatingElementRef: overlayRef})
+
+  useFocusZone({containerRef: overlayRef, disabled: !open || state !== 'listFocus'}, [position])
+  useFocusTrap({containerRef: overlayRef, disabled: !open || state !== 'listFocus'}, [position])
+
+  return (
+    <>
+      {renderAnchor({
+        ref: anchorRef,
+        id: anchorId,
+        'aria-labelledby': anchorId,
+        'aria-haspopup': 'listbox',
+        'aria-label': 'menu',
+        onClick: onAnchorClick,
+        onKeyDown: onAnchorKeyDown,
+        children: anchorContent,
+        tabIndex: 0
+      })}
+      {open ? (
+        <Overlay
+          initialFocusRef={anchorRef}
+          returnFocusRef={anchorRef}
+          onClickOutside={onDismiss}
+          onEscape={onDismiss}
+          ref={overlayRef}
+          {...position}
+        >
+          <List
+            {...listProps}
+            role="menu"
+            renderItem={({onClick, ...itemProps}) =>
+              renderItem({
+                ...itemProps,
+                role: 'menuitem',
+                onKeyPress: _event => {
+                  onAction && onAction(itemProps as ItemProps)
+                  onDismiss()
+                },
+                onClick: event => {
+                  onAction && onAction(itemProps as ItemProps)
+                  onClick && onClick(event)
```

Comment:
We could use the `?.` operator to make this a little cleaner :) 
```suggestion
                onKeyPress: _event => {
                  onAction?.(itemProps as ItemProps)
                  onDismiss()
                },
                onClick: event => {
                  onAction?.(itemProps as ItemProps)
                  onClick?.(event)
```